src/server/pkg/auth/handler.go

```diff
@@ -0,0 +1,141 @@
+package auth
+
+import (
+	"context"
+
+	"github.com/pachyderm/pachyderm/src/client/auth"
+
+	"google.golang.org/grpc"
+)
+
+type authHandlerFn func(*Interceptor) authHandler
+
+// authHandler methods return an error if the request is not permitted
+type authHandler interface {
+	unary(ctx context.Context, info *grpc.UnaryServerInfo, req interface{}) error
+	stream(ctx context.Context, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error
+	requireAuthEnabled()
+}
+
+func requireAuthEnabled(f authHandlerFn) authHandlerFn {
+	return func(i *Interceptor) authHandler {
+		h := f(i)
+		h.requireAuthEnabled()
```

Comment:
It seems like, if you wanted, you could remove `requireAuthEnabled` from the `authHandler` interface (and remove it as a field from the other handlers, so that each handler doesn't have to check if auth is enabled), and change the types a bit (this also uses the suggestion in the last comment of consolidating the interface):
```
// Calls 'WhoAmI' and then calls an innerAuthHandler
type authHandler func(context.Context, *grpc.UnaryServerInfo) error

// innerAuthHandler must be wrapped by 'requireAuthEnabled' or 'permitAuthDisabled'
type innerAuthHandler func(context.Context, *grpc.UnaryServerInfo, *auth.WhoAmIResponse) error
type authHandler func(*Interceptor) innerAuthHandler

func newAuthEnabledHandler(required bool, inner innerAuthHandler) authHandler {
	return func(ctx context.Context, info *grpc.UnaryServerInfo) {
		pachClient := h.interceptor.env.GetPachClient(ctx)
		resp, err := pachClient.WhoAmI(pachClient.Ctx(), &auth.WhoAmIRequest{})
		if err != nil {
			if auth.IsErrNotActivated(err) && !h.required {
				return nil
			}
			return err
		}
		return h.inner(ctx, info, resp)
	}
} 

func requireAuthEnabled(f authHandlerFn) authHandlerFn {
	return func(i *Interceptor) authHandler {
		return newAuthEnabledHandler(/* auth required: */ true, f(i))
	}
}

func permitAuthDisabled(f authHandlerFn) authHandlerFn {
	return func(i *Interceptor) authHandler {
		return newAuthEnabledHandler(/* auth required: */ false, f(i))
	}
}
...
...
// The other handlers are mostly reduced to 'isPermitted' (which could probably just be defined inline, if you wanted)
func authenticated(i *Interceptor) innerAuthHandler {
	a := &authenticatedHandler{
		interceptor: i,
	}
	return a.isPermitted
}
...
...
// this needs a wrapper now, but I think that's the only change
"/auth.API/Deactivate":               permitAuthDisabled(adminOnly),
...
...
// This is what the interceptors looks like afterwards:
func (i *Interceptor) InterceptUnary(...) (interface{}, error) {
	...
	if err := a(i)(ctx, info); err != nil {
	...
}

func (i *Interceptor) InterceptStream(...) (interface{}, error) {
	...
	if err := a(i)(ctx, info); err != nil {
	...
}
```

I think that would get you the current behavior, and the other handlers could know whether auth is enabled based on whether `who` is nil (or, if always wrapped in `requireAuthEnabled()`, assume `who` is never nil and assume auth is on).

Reply:
as an aside, what do you think of the names `authEnabledAnd` and `authDisabledOr`, and then `allowed`, `authenticated` and `admin` for the inner auth handlers (the idea is `enabled` means we don't need a variadic argumement—there's always an inner handler—and the table looks like:
```
....
"/auth.API/Activate":            authDisabledOr(allowed),
....
"/auth.API/GetOneTimePassword":  authEnabledAnd(authenticated),
...
"/auth.API/Deactivate":          authDisabledOr(admin),
"/auth.API/SetConfiguration":    authEnabledAnd(admin),
...
"/debug.Debug/Profile":          authDisabledOr(admin),
```

It means extra typing in the table, because all of the non-auth APIs become `authDisabledOr`, like:
```
"/pfs.API/CreateRepo":      authDisabledOr(authenticated),
"/pfs.API/InspectRepo":     authDisabledOr(authenticated),
"/pfs.API/ListRepo":        authDisabledOr(authenticated),
"/pfs.API/DeleteRepo":      authDisabledOr(authenticated),
"/pfs.API/StartCommit":     authDisabledOr(authenticated),
...
```

so it's maybe a little more repetitive, but I thought it might read nicely. Anyway, just an idea)